src/Features/Core/Portable/ValidateRegexString/AbstractValidateRegexStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,283 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.RegularExpressions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ValidateRegexString
+{
+    internal abstract class AbstractValidateRegexStringDiagnosticAnalyzer<TSyntaxKind>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TSyntaxKind : struct
+    {
+        private const string _patternName = "pattern";
+        private readonly int _stringLiteralKind;
+
+        protected AbstractValidateRegexStringDiagnosticAnalyzer(int stringLiteralKind)
+            : base(IDEDiagnosticIds.RegexPatternDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Regex_issue_0), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _stringLiteralKind = stringLiteralKind;
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract IVirtualCharService GetVirtualCharService();
+
+        protected abstract IParameterSymbol DetermineParameter(SemanticModel semanticModel, SyntaxNode argumentNode, CancellationToken cancellationToken);
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSemanticModelAction(AnalyzeSemanticModel);
```

Comment:
I did this as a semantic model analyzer so that we wouldn't need callbacks and processing for every string in the world.  Instead, we could process a file at a time, checking the user options once, and then doing a full pass to check each string in it.  @sharwell let me know if there's a preference on the best way to design these to be good perf citizens.